interfaces/acados_template/acados_template/acados_casadi_ocp_solver.py

```diff
@@ -248,75 +296,177 @@ def get_dim_flat(self, field: str):
+            return lam.flatten()
+        elif field in ['sl', 'su', 'z']:
+            return np.empty((0, 1))  # Only empty is supported for now. TODO: extend.
         else:
             raise NotImplementedError(f"Field '{field}' is not implemented in AcadosCasadiOcpSolver")
 
     def get_flat(self, field_: str) -> np.ndarray:
         """
         Get concatenation of all stages of last solution of the solver.
 
-        :param field: string in ['x', 'u', 'z', 'pi', 'lam', 'sl', 'su', 'p', 'p_global']
+        :param field: string in ['x', 'u', 'pi', 'lam', 'z', 'sl', 'su']
 
         .. note:: The parameter 'p_global' has no stage-wise structure and is processed in a memory saving manner by default. \n
                 In order to read the 'p_global' parameter, the option 'save_p_global' must be set to 'True' upon instantiation. \n
         """
-        raise NotImplementedError()
-
+        if self.nlp_sol is None:
+            raise ValueError('No solution available. Please call solve() first.')
+        dims = self.acados_ocp.dims
+        result = []
+
+        if field_ in ['x', 'lam', 'sl', 'su']:
+            for i in range(dims.N+1):
+                result.append(self.get(i, field_))
+            return np.concatenate(result)
+        elif field_ in ['u', 'pi', 'z']:
+            for i in range(dims.N):
+                result.append(self.get(i, field_))
+            return np.concatenate(result)
+        # casadi variables. TODO: maybe remove this.
+        elif field_ == 'lam_x':
+            return self.nlp_sol_lam_x.flatten()
+        elif field_ == 'lam_g':
+            return self.nlp_sol_lam_g.flatten()
+        elif field_ == 'lam_p':
+            return self.nlp_sol['lam_p'].full().flatten()
+        else:
+            raise NotImplementedError(f"Field '{field_}' is not implemented in get_flat().")
 
     def set_flat(self, field_: str, value_: np.ndarray) -> None:
         """
         Set concatenation solver initialization.
 
-        :param field: string in ['x', 'u', 'z', 'pi', 'lam', 'sl', 'su', 'p']
+        :param field: string in ['x', 'u', 'lam', pi]
         """
-        raise NotImplementedError()
-
+        dims = self.acados_ocp.dims
+        if field_ == 'x':
+            for i in range(dims.N+1):
+                self.set(i, 'x', value_[i*dims.nx:(i+1)*dims.nx])
+        elif field_ == 'u':
+            for i in range(dims.N):
+                self.set(i, 'u', value_[i*dims.nu:(i+1)*dims.nu])
+        elif field_ == 'pi':
+            for i in range(dims.N):
+                self.set(i, 'pi', value_[i*dims.nx:(i+1)*dims.nx])
+        elif field_ == 'lam':
+            offset = 0
+            for i in range(dims.N+1):
+                if i == 0:
+                    bx_length = self.acados_ocp.constraints.lbx_0.size
+                    bu_length = self.acados_ocp.constraints.lbu.size
+                    h_length = self.acados_ocp.constraints.lh_0.size + self.acados_ocp.constraints.lphi_0.size
```

Comment:
Please refactor here to use dims, "nbu" etc, as I did in the other functions.
Note that `h_length` should be replaced by `ng+nh+nphi`, as I tried to make clear renaming the `index_map` entry.